master/internal/api_runs.go

```diff
@@ -911,3 +914,60 @@ func pauseResumeAction(ctx context.Context, isPause bool, projectID int32,
 	}
 	return results, nil
 }
+
+// GetRunMetadata returns the metadata of a run.
+func (a *apiServer) GetRunMetadata(
+	ctx context.Context, req *apiv1.GetRunMetadataRequest,
+) (*apiv1.GetRunMetadataResponse, error) {
+	// TODO(runs) run specific RBAC.
+	if err := trials.CanGetTrialsExperimentAndCheckCanDoAction(ctx, int(req.RunId),
+		experiment.AuthZProvider.Get().CanGetExperimentArtifacts); err != nil {
+		return nil, err
+	}
+	metadata, err := db.GetRunMetadata(ctx, int(req.RunId))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.GetRunMetadataResponse{
+		Metadata: protoutils.ToStruct(metadata),
+	}, nil
+}
+
+// PostRunMetadata updates the metadata of a run.
+func (a *apiServer) PostRunMetadata(
+	ctx context.Context, req *apiv1.PostRunMetadataRequest,
+) (*apiv1.PostRunMetadataResponse, error) {
+	// TODO(runs) run specific RBAC.
+	if err := trials.CanGetTrialsExperimentAndCheckCanDoAction(ctx, int(req.RunId),
+		experiment.AuthZProvider.Get().CanEditExperiment); err != nil {
+		return nil, err
+	}
+
+	if len(req.Metadata.AsMap()) == 0 || req.Metadata.AsMap() == nil {
+		return nil, status.Error(codes.InvalidArgument, "metadata cannot be empty")
+	}
+
+	// Flatten Request Metadata.
+	flatMetadata, err := run.FlattenMetadata(req.Metadata.AsMap())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	flatKeys := make(map[string]struct{})
+	for _, record := range flatMetadata {
+		flatKeys[record.FlatKey] = struct{}{}
+	}
+
+	// Update the metadata in DB
+	result, err := db.UpdateRunMetadata(ctx,
+		int(req.RunId),
+		req.Metadata.AsMap(),
+		flatMetadata,
+	)
+	if err != nil && status.Code(err) == codes.InvalidArgument {
```

Comment:
Why are we checking that the error code is in invalid argument here?

Reply:
Good catch! I believe this had been carry over from the removal of the merge logic, will remove!